File: desktop/macos/Desktop/Sources/Bluetooth/Connections/BeeDeviceConnection.swift
@preconcurrency import CoreBluetooth
import Foundation
import os.log

/// Device connection implementation for Bee devices
/// Uses AAC audio codec with ADTS framing
/// Ported from: omi/app/lib/services/devices/bee_connection.dart
final class BeeDeviceConnection: BaseDeviceConnection {

  // MARK: - Constants

  private static let controlCharacteristicUUID = "05e1f93c-d8d0-5ed8-dd88-379e4c1a3e3e"
  private static let audioCharacteristicUUID = "b189a505-a86c-11ee-a5fb-8f2089a49e7e"

  private enum Command {
    static let mute: UInt16 = 0xC006
    static let unmute: UInt16 = 0xC006
    static let battery: UInt16 = 0xC00F
  }

  // MARK: - Private Properties

  private let logger = Logger(subsystem: "me.omi.desktop", category: "BeeDeviceConnection")

  private var audioBuffer = [UInt8]()
  private var controlSubscription: Task<Void, Never>?
  private var audioSubscription: Task<Void, Never>?
  private let responseOperations: DeviceOperationBroker<UInt16, Data>
  private let responseGate = UncorrelatedOperationGate<UInt16>()
  private let commandQueue = DeviceCommandQueue()
  private var audioStartAttempted = false
  private lazy var audioStreamController = DeviceAudioStreamController(
    start: { [weak self] in
      guard let self else { throw DeviceTransportError.disposed }
      try await self.startAudioSession()
    },
    stop: { [weak self] in
      guard let self else { return }
      try await self.stopAudioSession()
    }
  )

  // MARK: - Initialization

  override init(
    device: BtDevice,
    transport: DeviceTransport,
    operationClock: any DeviceOperationClock = ContinuousDeviceOperationClock()
  ) {
    self.responseOperations = DeviceOperationBroker(clock: operationClock)
    super.init(
      device: device,
      transport: transport,
      operationClock: operationClock
    )
  }

  // MARK: - Connection

  override func prepareDeviceAfterConnect() async throws {
    try await operationClock.sleep(for: .seconds(1))

    startControlListener()
    startAudioListener()
  }

  override func teardownDevice() async {
    await audioStreamController.finish()
    await commandQueue.close()
    audioBuffer.removeAll(keepingCapacity: false)

    controlSubscription?.cancel()
    audioSubscription?.cancel()
    controlSubscription = nil
    audioSubscription = nil
    await responseOperations.cancelAll(reason: .disconnected)
    responseGate.reset()
  }

  // MARK: - Listeners

  private func startControlListener() {
    let stream = transport.getCharacteristicStream(
      serviceUUID: DeviceUUIDs.Bee.service,
      characteristicUUID: CBUUID(string: Self.controlCharacteristicUUID)
    )

    controlSubscription = Task { [weak self] in
      do {
        for try await data in stream {
          self?.handleControlResponse(Array(data))
        }
      } catch {
        self?.logger.debug("Control stream ended")
      }
    }
  }

  private func startAudioListener() {
    let stream = transport.getCharacteristicStream(
      serviceUUID: DeviceUUIDs.Bee.service,
      characteristicUUID: CBUUID(string: Self.audioCharacteristicUUID)
    )

    audioSubscription = Task { [weak self] in
      do {
        for try await data in stream {
          if let frame = self?.processAudioPacket(Array(data)) {
            self?.audioStreamController.yield(Data(frame))
          }
        }
      } catch {
        self?.logger.debug("Audio stream ended")
      }
    }
  }

  // MARK: - Response Handling

  /// Parses one physical control notification. Internal for deterministic
  /// command-correlation tests without Bluetooth hardware.
  func handleControlResponse(_ data: [UInt8]) {
    guard data.count >= 2 else { return }

    let responseCode = UInt16(data[0]) | (UInt16(data[1]) << 8)
    let payload = data.count > 2 ? Array(data.dropFirst(2)) : []

    // Check for echo response (0x8000)
    if responseCode == 0x8000 && payload.count >= 2 {
      let echoedCmd = UInt16(payload[0]) | (UInt16(payload[1]) << 8)
      let actualPayload = payload.count > 2 ? Array(payload.dropFirst(2)) : []

      completeResponse(commandID: echoedCmd, payload: actualPayload)
    } else {
      completeResponse(commandID: responseCode, payload: payload)
    }
  }

  private func completeResponse(commandID: UInt16, payload: [UInt8]) {
    guard let handle = responseGate.takeHandleForCallback(key: commandID) else { return }
    Task {
      await responseOperations.succeed(
        handle: handle,
        value: Data(payload)
      )
    }
  }

  // MARK: - Audio Processing

  /// Process AAC audio packet with ADTS framing
  func processAudioPacket(_ data: [UInt8]) -> [UInt8]? {
    guard data.count >= 2 else { return nil }

    audioBuffer.append(contentsOf: data.dropFirst(2))
    return Self.nextADTSFrame(from: &audioBuffer)
  }

  /// Pop the next complete ADTS/AAC frame from `buffer`, skipping leading
  /// non-sync or malformed-header bytes. Returns nil when no complete valid
  /// frame is available yet.
  ///
  /// A valid ADTS frame is at least 7 bytes (the header length is part of the
  /// frameLength field). A decoded `frameLength < 7` — e.g. a false sync where
  /// 0xFF 0xFx matched inside AAC payload after a dropped BLE notification — was
  /// previously accepted: `buffer.count >= 0` is always true, `prefix(0)` is
  /// empty, and `removeFirst(0)` advances nothing, so the corrupt header stayed
  /// at the head and every later packet re-hit it — audio went permanently
  /// silent while the buffer grew without bound. Treat `< 7` as a false sync and
  /// advance one byte so the scanner keeps making progress.
  nonisolated static func nextADTSFrame(from buffer: inout [UInt8]) -> [UInt8]? {
    while buffer.count >= 7 {
      // Look for ADTS sync word (0xFF 0xFx)
      guard buffer[0] == 0xFF, (buffer[1] & 0xF0) == 0xF0 else {
        buffer.removeFirst()
        continue
      }

      // Extract frame length from ADTS header
      let frameLength = (Int(buffer[3] & 0x03) << 11) | (Int(buffer[4]) << 3) | (Int(buffer[5] & 0xE0) >> 5)

      guard frameLength >= 7 else {
        // False sync: 0xFF 0xFx matched but the header is invalid.
        buffer.removeFirst()
        continue
      }

      if buffer.count >= frameLength {
        let frame = Array(buffer.prefix(frameLength))
        buffer.removeFirst(frameLength)
        return frame
      }
      break
    }
    return nil
  }

  // MARK: - Command Sending

  private func sendCommand(cmdId: UInt16, payload: [UInt8]) async throws -> [UInt8]? {
    try await commandQueue.run { [weak self] in
      guard let self else { throw DeviceTransportError.disposed }
      return try await self.sendCommandSerially(cmdId: cmdId, payload: payload)
    }
  }

  private func sendCommandSerially(cmdId: UInt16, payload: [UInt8]) async throws -> [UInt8]? {
    guard responseGate.canStart(cmdId) else {
      throw DeviceConnectionError.operationFailed(
        "A previous command response is ambiguous; reconnect the device"
      )
    }
    let command: [UInt8] = [UInt8(cmdId & 0xFF), UInt8((cmdId >> 8) & 0xFF)] + payload

    do {
      let response = try await responseOperations.perform(
        key: cmdId,
        timeout: .seconds(5),
        onTerminal: { [weak self] handle, termination in
          self?.responseGate.terminal(handle: handle, termination: termination)
        }
      ) { [weak self] handle in
        guard let self else { throw DeviceTransportError.disposed }
        guard self.responseGate.register(handle) else {
          throw DeviceConnectionError.operationFailed(
            "Command callback identity is no longer trustworthy"
          )
        }
        try await self.transport.writeCharacteristic(
          data: Data(command),
          serviceUUID: DeviceUUIDs.Bee.service,
          characteristicUUID: CBUUID(string: Self.controlCharacteristicUUID),
          withResponse: true
        )
      }
      return Array(response)
    } catch DeviceOperationBrokerError.timedOut {
      return nil
    }
  }

  // MARK: - Battery

  override func getBatteryLevel() async -> Int {
    guard await isConnected() else { return -1 }

    do {
      guard let response = try await sendCommand(cmdId: Command.battery, payload: []),
        response.count >= 2
      else {
        return -1
      }
      return Int(response[0])
    } catch {
      logger.debug("Error retrieving battery: \(error.localizedDescription)")
      return -1
    }
  }

  /// Get battery state including charging status
  func getBatteryState() async -> (level: Int, isCharging: Bool)? {
    guard await isConnected() else { return nil }

    do {
      guard let response = try await sendCommand(cmdId: Command.battery, payload: []),
        response.count >= 2
      else {
        return nil
      }
      return (level: Int(response[0]), isCharging: response[1] != 0)
    } catch {
      return nil
    }
  }

  override func getBatteryLevelStream() -> AsyncThrowingStream<Int, Error> {
    // Bee uses command-based battery retrieval, poll every 60 seconds
    return AsyncThrowingStream { [weak self] continuation in
      let pollingTask = Task { [weak self] in
        guard let self = self else {
          continuation.finish()
          return
        }

        var lastLevel: Int?

        // Get initial battery level
        let initialLevel = await self.getBatteryLevel()
        if initialLevel >= 0 {
          lastLevel = initialLevel
          continuation.yield(initialLevel)
        }

        // Poll every 60 seconds
        while await self.isConnected() {
          do {
            try await self.operationClock.sleep(for: .seconds(60))
          } catch {
            break
          }
          guard !Task.isCancelled else { break }

          let level = await self.getBatteryLevel()
          if level >= 0 && level != lastLevel {
            lastLevel = level
            continuation.yield(level)
          }
        }

        continuation.finish()
      }
      continuation.onTermination = { @Sendable _ in
        pollingTask.cancel()
      }
    }
  }

  // MARK: - Audio

  override func getAudioCodec() async -> BleAudioCodec {
    .aac
  }

  override func getAudioStream() -> AsyncThrowingStream<Data, Error> {
    audioStreamController.makeStream()
  }

  private func startAudioSession() async throws {
    audioBuffer.removeAll(keepingCapacity: true)
    audioStartAttempted = true
    guard try await sendCommand(cmdId: Command.unmute, payload: [0x01]) != nil else {
      throw DeviceConnectionError.operationFailed("Bee did not acknowledge audio start")
    }
  }

  private func stopAudioSession() async throws {
    // Notifications can still arrive while the mute acknowledgement is in
    // flight. Clear both before and after teardown so no partial AAC frame
    // can cross the physical recording-session boundary.
    defer { audioBuffer.removeAll(keepingCapacity: true) }
    audioBuffer.removeAll(keepingCapacity: true)
    guard audioStartAttempted else { return }
    audioStartAttempted = false

    if responseGate.canStart(Command.mute) {
      do {
        guard try await sendCommand(cmdId: Command.mute, payload: [0x00]) != nil else {
          throw DeviceConnectionError.operationFailed(
            "Bee did not acknowledge audio stop"
          )
        }
        return
      } catch {
        await transport.disconnect()
        throw error
      }
    }

    // Unmute and mute share one uncorrelated device command ID. If setup
    // was cancelled after the unmute write, its response key is poisoned;
    // issue the compensating mute directly on the still-valid BLE write
    // channel, then retire the physical session rather than reopening that
    // command ID and risking a late mute response aliasing a new unmute.
    let command = [
      UInt8(Command.mute & 0xFF),
      UInt8((Command.mute >> 8) & 0xFF),
      0x00,
    ]
    do {
      try await commandQueue.run { [weak self] in
        guard let self else { throw DeviceTransportError.disposed }
        try await self.transport.writeCharacteristic(
          data: Data(command),
          serviceUUID: DeviceUUIDs.Bee.service,
          characteristicUUID: CBUUID(string: Self.controlCharacteristicUUID),
          withResponse: true
        )
      }
    } catch {
      logger.debug("Compensating Bee mute could not be confirmed: \(error.localizedDescription)")
    }
    await transport.disconnect()
    throw DeviceConnectionError.operationFailed(
      "Bee audio command identity is ambiguous; reconnect the device"
    )
  }

  // MARK: - Unsupported Features

  override func getButtonState() async -> [UInt8] { [] }
  override func getButtonStream() -> AsyncThrowingStream<[UInt8], Error> {
    AsyncThrowingStream { $0.finish() }
  }
  override func hasPhotoStreaming() async -> Bool { false }
  override func getAccelerometerStream() -> AsyncThrowingStream<AccelerometerData, Error> {
    AsyncThrowingStream { $0.finish() }
  }
  override func getFeatures() async -> OmiFeatures { [] }

  // MARK: - Device Info

  override func updateDeviceInfo() async {
    device.modelNumber = "Bee"
    device.firmwareRevision = "1.0.0"
    device.hardwareRevision = "1.0.0"
    device.manufacturerName = "Bee"
  }
}

File: desktop/windows/src/renderer/src/device/connections/beeDeviceConnection.ts
/**
 * Bee connection — Windows port of macOS Connections/BeeDeviceConnection.swift.
 * Commands multiplex over one control characteristic as [u16 id LE, payload]
 * frames; mute and unmute share command id 0xC006, which is why a poisoned
 * mute identity retires the whole session instead of retrying.
 */

import { BEE_UUIDS } from '../protocol/uuids'
import type { BleAudioCodec } from '../protocol/deviceTypes'
import {
  DeviceOperationBroker,
  DeviceOperationBrokerError,
  UncorrelatedOperationGate
} from '../session/deviceOperationBroker'
import { DeviceCommandQueue } from '../session/deviceCommandQueue'
import { DeviceAudioStreamController } from '../session/deviceAudioStreamController'
import {
  BaseDeviceConnection,
  DeviceConnectionError,
  finishedStream,
  type AccelerometerData,
  type StreamSubscriber,
  type StreamSubscription
} from './deviceConnection'
import { AdtsFrameExtractor } from './adtsFrameExtractor'
import type { DeviceTransport } from '../transport/deviceTransport'
import type { BtDevice } from '../protocol/btDevice'
import type { DeviceOperationClock } from '../session/deviceOperationBroker'

/** Mute and unmute share this id; battery is 0xC00F. */
const CMD_AUDIO_MUTE_UNMUTE = 0xc006
const CMD_BATTERY = 0xc00f
const ECHO_RESPONSE_CODE = 0x8000
const COMMAND_TIMEOUT_MS = 5_000
const POST_CONNECT_SETTLE_MS = 1_000
const BATTERY_POLL_INTERVAL_MS = 60_000
const AUDIO_PACKET_HEADER_BYTES = 2

const commandKey = (cmdId: number): string => `bee-cmd:${cmdId}`

export interface BeeBatteryState {
  level: number
  isCharging: boolean
}

export class BeeDeviceConnection extends BaseDeviceConnection {
  private readonly commandQueue = new DeviceCommandQueue()
  private readonly commandBroker: DeviceOperationBroker
  private readonly commandGate = new UncorrelatedOperationGate()
  private readonly adts = new AdtsFrameExtractor()
  private readonly audioController: DeviceAudioStreamController

  constructor(args: {
    device: BtDevice
    transport: DeviceTransport
    clock?: DeviceOperationClock
  }) {
    super(args)
    this.commandBroker = new DeviceOperationBroker(this.clock)
    this.audioController = new DeviceAudioStreamController({
      start: () => this.startAudioSession(),
      stop: () => this.stopAudioSession()
    })
  }

  protected override async updateDeviceInfo(): Promise<void> {
    this.device.modelNumber = 'Bee'
    this.device.firmwareRevision = '1.0.0'
    this.device.hardwareRevision = '1.0.0'
    this.device.manufacturerName = 'Bee'
  }

  protected override async prepareDeviceAfterConnect(): Promise<void> {
    await this.settle(POST_CONNECT_SETTLE_MS)
    this.ensureLifecycleIsActive()
    this.transport.subscribeCharacteristic(BEE_UUIDS.service, BEE_UUIDS.control, {
      onData: (data) => this.handleControlResponse(data),
      onFinish: () => undefined
    })
    this.transport.subscribeCharacteristic(BEE_UUIDS.service, BEE_UUIDS.audio, {
      onData: (data) => this.handleAudioPacket(data),
      onFinish: (error) => {
        void this.audioController.finish(error)
      }
    })
  }

  protected override async teardownDevice(): Promise<void> {
    await this.audioController.finish(null)
    this.commandBroker.cancelAll('disconnected')
    this.commandGate.reset()
    await this.commandQueue.close()
  }

  // --- command channel -------------------------------------------------------

  /** Sends one command frame and awaits its correlated response; a timeout
   *  resolves null (matching mac), a poisoned identity throws. */
  private sendCommand(cmdId: number, payload: Uint8Array): Promise<Uint8Array | null> {
    return this.commandQueue.run(async () => {
      const key = commandKey(cmdId)
      const handle = this.commandGate.register(key)
      if (handle === null) {
        throw DeviceConnectionError.operationFailed(
          'A previous command response is ambiguous; reconnect the device'
        )
      }
      const frame = new Uint8Array(2 + payload.length)
      frame[0] = cmdId & 0xff
      frame[1] = (cmdId >> 8) & 0xff
      frame.set(payload, 2)
      try {
        return await this.commandBroker.perform<Uint8Array>({
          key,
          timeoutMs: COMMAND_TIMEOUT_MS,
          onTerminal: (termination) => this.commandGate.terminal(handle, termination),
          start: () =>
            this.transport.writeCharacteristic({
              serviceUuid: BEE_UUIDS.service,
              characteristicUuid: BEE_UUIDS.control,
              data: frame,
              withResponse: true
            })
        })
      } catch (error) {
        if (error instanceof DeviceOperationBrokerError && error.kind === 'timedOut') {
          return null
        }
        throw this.mapTransportError(error)
      }
    })
  }

  private handleControlResponse(data: Uint8Array): void {
    if (data.length < 2) return
    const responseCode = data[0] | (data[1] << 8)
    const payload = data.subarray(2)
    if (responseCode === ECHO_RESPONSE_CODE && payload.length >= 2) {
      const echoedCmd = payload[0] | (payload[1] << 8)
      this.completeCommand(echoedCmd, payload.subarray(2))
      return
    }
    this.completeCommand(responseCode, payload)
  }

  private completeCommand(cmdId: number, payload: Uint8Array): void {
    const key = commandKey(cmdId)
    if (this.commandGate.takeHandleForCallback(key) === null) return
    this.commandBroker.succeed(key, payload.slice())
  }

  // --- audio -----------------------------------------------------------------

  private handleAudioPacket(data: Uint8Array): void {
    if (data.length <= AUDIO_PACKET_HEADER_BYTES) return
    const frames = this.adts.push(data.subarray(AUDIO_PACKET_HEADER_BYTES))
    for (const frame of frames) {
      this.audioController.yieldFrame(frame)
    }
  }

  private async startAudioSession(): Promise<void> {
    this.adts.clear()
    const response = await this.sendCommand(CMD_AUDIO_MUTE_UNMUTE, Uint8Array.from([0x01]))
    if (response === null) {
      throw DeviceConnectionError.operationFailed('Bee did not acknowledge audio start')
    }
  }

  private async stopAudioSession(): Promise<void> {
    this.adts.clear()
    // A mute command on a dead link just burns the 5 second command timeout
    // before teardown can continue, delaying the disconnect notification and
    // the reconnect that follows it.
    if (!this.transport.isConnected()) {
      this.commandBroker.cancelAll('disconnected')
      return
    }
    try {
      if (this.commandGate.isPoisoned(commandKey(CMD_AUDIO_MUTE_UNMUTE))) {
        // Mute and unmute share one command id: with the identity poisoned, a
        // late response could alias a future unmute, so write the raw mute
        // uncorrelated and retire the session.
        await this.commandQueue.run(() =>
          this.transport.writeCharacteristic({
            serviceUuid: BEE_UUIDS.service,
            characteristicUuid: BEE_UUIDS.control,
            data: Uint8Array.from([0x06, 0xc0, 0x00]),
            withResponse: true
          })
        )
        await this.transport.disconnect()
        throw DeviceConnectionError.operationFailed(
          'Bee audio command identity is ambiguous; reconnect the device'
        )
      }
      const response = await this.sendCommand(CMD_AUDIO_MUTE_UNMUTE, Uint8Array.from([0x00]))
      if (response === null) {
        throw DeviceConnectionError.operationFailed('Bee did not acknowledge audio stop')
      }
    } catch (error) {
      await this.transport.disconnect()
      throw error
    } finally {
      this.adts.clear()
    }
  }

  override async getAudioCodec(): Promise<BleAudioCodec> {
    return 'aac'
  }

  override getAudioStream(subscriber: StreamSubscriber<Uint8Array>): StreamSubscription {
    const subscription = this.audioController.subscribe({
      onFrame: (frame) => subscriber.onValue(frame),
      onFinish: (error) => subscriber.onFinish(error)
    })
    return { cancel: () => subscription.cancel() }
  }

  // --- battery ---------------------------------------------------------------

  async getBatteryState(): Promise<BeeBatteryState | null> {
    try {
      const response = await this.sendCommand(CMD_BATTERY, new Uint8Array(0))
      if (response === null || response.length < 2) return null
      return { level: response[0], isCharging: response[1] !== 0 }
    } catch {
      return null
    }
  }

  override async getBatteryLevel(): Promise<number> {
    const state = await this.getBatteryState()
    return state?.level ?? -1
  }

  override getBatteryLevelStream(subscriber: StreamSubscriber<number>): StreamSubscription {
    const abort = new AbortController()
    this.sessionAbort.signal.addEventListener('abort', () => abort.abort(), { once: true })
    void (async () => {
      let lastLevel: number | null = null
      for (;;) {
        const state = await this.getBatteryState()
        if (abort.signal.aborted) return
        if (state !== null && state.level !== lastLevel) {
          lastLevel = state.level
          subscriber.onValue(state.level)
        }
        const outcome = await this.clock.sleep(BATTERY_POLL_INTERVAL_MS, abort.signal)
        if (outcome !== 'elapsed') return
      }
    })()
    return { cancel: () => abort.abort() }
  }

  // --- unsupported surfaces --------------------------------------------------

  override async getButtonState(): Promise<Uint8Array> {
    return new Uint8Array(0)
  }

  override getButtonStream(subscriber: StreamSubscriber<Uint8Array>): StreamSubscription {
    return finishedStream(subscriber)
  }

  override getAccelerometerStream(
    subscriber: StreamSubscriber<AccelerometerData>
  ): StreamSubscription {
    return finishedStream(subscriber)
  }

  override async getFeatures(): Promise<[]> {
    return []
  }
}
